python/tk_multi_workfiles/work_area.py

```diff
@@ -264,54 +278,39 @@ def _get_settings_for_context(self, context, templates_to_find, settings_to_find
+        app_settings = WorkArea._settings_cache.get(context)
+
+        if app_settings is None:
             try:
                 # find settings for all instances of app in the environment picked for the given context:
-                other_settings = sgtk.platform.find_app_settings(app.engine.name, app.name, app.sgtk, context)
-                if not other_settings:
-                    # for backwards compatibility, look for settings for the 'tk-multi-workfiles' app as well
-                    other_settings = sgtk.platform.find_app_settings(app.engine.name, "tk-multi-workfiles", 
-                                                                     app.sgtk, context)
+                app_settings = sgtk.platform.find_app_settings(
+                    app.engine.name, app.name, app.sgtk, context, app.engine.instance_name
+                )
             finally:
-                # make sure the cache is updated:
-                WorkArea._settings_cache.add(app.engine.name, app.name, context, other_settings or {})
+                # Ignore any errors while looking for the settings
+                WorkArea._settings_cache.add(context, app_settings or {})
 
-        if len(other_settings) == 1:
-            return other_settings[0].get("settings")
-    
-        settings_by_engine = {}
-        for settings in other_settings:
-            settings_by_engine.setdefault(settings.get("engine_instance"), list()).append(settings)
-        
-        # can't handle more than one engine!  
-        if len(settings_by_engine) != 1:
-            return
-            
-        # ok, so have a single engine but multiple apps lets try 
-        # to find an app with the same instance name.
-
-        # first, get the instance name of the current app - we
-        # have to look through all engine apps to find this as
-        # the app itself doesn't know
-        app_instance_name = None
-        for instance_name, engine_app in app.engine.apps.iteritems():
-            if engine_app == app:
-                app_instance_name = instance_name
-                break
-        if not app_instance_name:
-            return
-    
-        # now look for settings for this specific instance in the engine settings:
-        for engine_name, engine_settings in settings_by_engine.iteritems():
-            for settings in engine_settings:
-                if settings.get("app_instance") == app_instance_name:
-                    return settings.get("settings")
+        if len(app_settings) == 1:
+            return app_settings[0].get("settings")
+
+        # There's more than one instance of that app for the engine instance, so we'll
+        # need to deterministically pick one. We'll pick the one with the same
+        # application instance name as the current app instance.
+        for settings in app_settings:
+            if settings.get("app_instance") == app.instance_name:
+                return settings.get("settings")
+
+        return None
```

Comment:
Would it make more sense at this point to log a warning and then pick the first instance found? So we try to provide the user with something half-sensible but also warn about the state of things?